Designate configuration-specific parameters based on PypeIt metadata

tbowers7 opened the pull request:
### PR Description

#### Initial Description

As described in #1958 , configuration-specific parameters had been polled from FITS headers directly.  This overrides edits to the ``.pypeit`` files.  As discussed in the developer telecon on 25 August 2025, the primacy of the Pypeit Reduction File as the source for TRUTH in a reduction should be maintained.  This PR aims to restore this primacy.

Of the calls to ``.config_specific_par()``, some really do need to send the raw data file / header rather than the metadata table (*e.g.* from `sensfunc`, where no metadata table exists).  As best I can figure, only the calls from ``inputfiles.py`` and ``chk_for_calibs.py`` are under conditions where the metadata table (from, *e.g.*, the ``.pypeit`` file) is required to possibly correct for bad FITS headers.

The solution was to check for the *type* of the input to the method, where the callers that need to rely upon editable metadata can send a row of the metadata `Table` and others can continue to send a filename or `HDUList` or whatever.  The method extracts the needed value as appropriate given the *type* of the input and sets and returns parameters as before.

The bulk of the work was to propagate these ``config_specific_par()`` changes to **ALL** relevant spectrographs!  In the process, I also did some import cleanup / standardization in the affected spectrograph classes.  In instruments with multiple possible configurations, ``if..elif..elif..else`` syntaxes were moved to a ``match..case..case`` syntax both for modernization and for clarity for the reader.

Finally, a unit test was added to ensure the functionality of both *type*s of input and that when changing metadata keys used to sort configurations, the ``config_specific_par`` actually changes.

#### Additions to the PR made along the way:

It is important to be sure metadata keys needed for ``config_specific_par()`` are actually in the PypeIt Reduction File or are configuration keys.  Various keys added to various spectrographs.

Added companion DevSuite PR (https://github.com/pypeit/PypeIt-development-suite/pull/369) that ~~updates the canned ``.pypeit`` files in the DevSuite to include the new columns written to satisfy ``config_specific_par()`` when passed a metadata table row~~ fixes some Gemini/GMOS MJD issues.

A couple issues were stinkers. The first was Python's dynamic typing and Bok/B&C not following the usual scheme of "300GPM" or "300L" for grating names... just "300" -- ``int(300)`` is not the same as ``str(300)``. The second was older Gemini/GMOS .pypeit files using observing epoch in the MJD metadata column... which makes for Bad Things when trying to use that value as an actual MJD...

See the "resolved comments" in the conversation below for a blow-by-blow of additional changes.

----

### ~~Secondary Change~~

~~Also, because I was cleaning imports in various places, remove ``__UTC__`` python <= 3.10 compatibility.  Since we now only support Python >= 3.11, remove these crutches left over from supporting 3.10.~~

~~These can be pulled out into a separate PR, if desired -- it is entirely contained in 52c94fa9067a0796b0523bb14dceef89ec7850d6.~~

This was merged into `develop` with PR #1994 

----

### Release Notes

Release notes ~~will be added after 1.18 is released and I merge in the `develop` at that time~~ have been added.

tbowers7 committed:
Remove ``__UTC__`` python <= 3.10 compatability

Since we now only support Python >= 3.11, remove these crutches left
over from supporting 3.10.

Also, clean up import order somewhat for sanity.

	modified:   pypeit/calibrations.py
	modified:   pypeit/inputfiles.py
	modified:   pypeit/pypeit.py
	modified:   pypeit/pypmsgs.py
	modified:   pypeit/scripts/chk_for_calibs.py
	modified:   pypeit/setup_gui/controller.py

tbowers7 committed:
Getting a solution working for LDT/DeVeny

Of the calls to ``.config_specific_par()``, some really do need to send
the raw data file / header rather than the metadata table.  As best I
can figure, only the calls from ``inputfiles.py`` and ``chk_for_calibs.py``
are under conditions where the metadata table (from, *e.g.*, the PypeIt
file) is required to possibly correct for bad FITS headers.

The next step will be to propagate these changes to ALL the other
spectrograph classes.

	modified:   pypeit/inputfiles.py
	modified:   pypeit/scripts/chk_for_calibs.py
	modified:   pypeit/spectrographs/ldt_deveny.py
	modified:   pypeit/spectrographs/spectrograph.py

tbowers7 committed:
Propagate ``config_specific_par()`` changes

... to ALL relevant spectrographs!

Also did some import cleanup / standardization in the affected
spectrograph classes.

	modified:   pypeit/spectrographs/aat_uhrf.py
	modified:   pypeit/spectrographs/apf_levy.py
	modified:   pypeit/spectrographs/bok_bc.py
	modified:   pypeit/spectrographs/gemini_flamingos.py
	modified:   pypeit/spectrographs/gemini_gmos.py
	modified:   pypeit/spectrographs/gemini_gnirs.py
	modified:   pypeit/spectrographs/gtc_osiris.py
	modified:   pypeit/spectrographs/keck_deimos.py
	modified:   pypeit/spectrographs/keck_hires.py
	modified:   pypeit/spectrographs/keck_kcwi.py
	modified:   pypeit/spectrographs/keck_lris.py
	modified:   pypeit/spectrographs/keck_mosfire.py
	modified:   pypeit/spectrographs/keck_nirspec.py
	modified:   pypeit/spectrographs/lbt_mods.py
	modified:   pypeit/spectrographs/ldt_deveny.py
	modified:   pypeit/spectrographs/mmt_binospec.py
	modified:   pypeit/spectrographs/mmt_bluechannel.py
	modified:   pypeit/spectrographs/mmt_mmirs.py
	modified:   pypeit/spectrographs/not_alfosc.py
	modified:   pypeit/spectrographs/ntt_efosc2.py
	modified:   pypeit/spectrographs/p200_dbsp.py
	modified:   pypeit/spectrographs/shane_kast.py
	modified:   pypeit/spectrographs/soar_goodman.py
	modified:   pypeit/spectrographs/spectrograph.py
	modified:   pypeit/spectrographs/subaru_focas.py
	modified:   pypeit/spectrographs/tng_dolores.py
	modified:   pypeit/spectrographs/vlt_fors.py
	modified:   pypeit/spectrographs/wht_isis.py

tbowers7 committed:
Merge branch 'develop' into issue_1958

tbowers7 committed:
Add unit test for ``config_specific_par()``

Tests both input methods on ``shane_kast_blue`` and checks that the
config-specific paramater for that instrument does, indeed, change when
the configuration changes.

	modified:   pypeit/tests/test_spectrographs.py

tbowers7 committed:
Merge branch 'develop' into issue_1958

tbowers7 committed:
Fixes to deal with failed tests

Yes, the initial round of testing went poorly.

--- PYTEST PYPEIT UNIT TESTS PASSED  261 passed, 445 warnings in 541.44s (0:09:01) ---
--- PYTEST UNIT TESTS FAILED  3 failed, 154 passed, 126104 warnings in 1040.84s (0:17:20) ---
--- PYTEST VET TESTS FAILED  20 failed, 47 passed, 142 warnings in 4640.21s (1:17:20) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 52/255 TESTS  ---

This commit addresses most of the errors found when going through the
report, but there will still be some errors.  In the next cloud run,
I will be copying out all the "*.log" files to be able to see more than
the last line of the log for what happened.

Of these fixes, most have to do with making sure metadata keys needed
for ``config_specific_par()`` are actually in the PypeIt Reduction File
or are configuration keys.

The NOT/NTT errors were related to "#" characters in the grating name
that get scrubbed by the metadata class.  And a couple of things I just
plain missed the first time through.

Here's hoping!  🤞

	modified:   pypeit/scripts/chk_for_calibs.py
	modified:   pypeit/spectrographs/gemini_gnirs.py
	modified:   pypeit/spectrographs/gtc_osiris.py
	modified:   pypeit/spectrographs/keck_deimos.py
	modified:   pypeit/spectrographs/keck_lris.py
	modified:   pypeit/spectrographs/keck_mosfire.py
	modified:   pypeit/spectrographs/ldt_deveny.py
	modified:   pypeit/spectrographs/mmt_bluechannel.py
	modified:   pypeit/spectrographs/not_alfosc.py
	modified:   pypeit/spectrographs/ntt_efosc2.py
	modified:   pypeit/spectrographs/shane_kast.py

tbowers7 commented:
```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  261 passed, 445 warnings in 541.44s (0:09:01) ---
--- PYTEST UNIT TESTS FAILED  3 failed, 154 passed, 126104 warnings in 1040.84s (0:17:20) ---
--- PYTEST VET TESTS FAILED  20 failed, 47 passed, 142 warnings in 4640.21s (1:17:20) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 52/255 TESTS  ---
Failed tests:
...
Skipped tests:
...
Testing Started at 2025-08-27T21:46:34.279774
Testing Completed at 2025-08-28T10:17:37.533897
Total Time: 12:31:03.254123
```

Well, the first round of testing didn't go quite as well as I had hoped.

Made changes to address most of the errors found when going through the report, but there will still be some errors (MODS, in particular, but I'm sure I didn't completely address all the errors from the first round here).  In the currently-processing cloud run, I will be copying out all the "``*.log``" files to be able to see more than the last line of the log for what happened.

Of the fixes, most have to do with making sure metadata keys needed for ``config_specific_par()`` are actually in the PypeIt Reduction File or are configuration keys.

The NOT/NTT errors were related to "``#``" characters in the grating name that get scrubbed by the metadata class.  And a couple of things I just plain missed the first time through.

Here's hoping!  🤞

tbowers7 committed:
Operational and typographic fixes

Fold the recent APF/Levy binning update into this scheme, fix the
Keck/MOSFIRE duplicate column and missed typos.

These should be the last changes here needed to make all tests pass.
What's next is to propagate the new ``configuration_keys()`` columns
into the ``.pypeit`` files in the DevSuite that are used for the
instrument tests.  This mismatch between older PypeIt Reduction Files
and what is expected here caused most of the failues in the last run.

--- PYTEST PYPEIT UNIT TESTS PASSED  261 passed, 441 warnings in 414.78s (0:06:54) ---
--- PYTEST UNIT TESTS FAILED  5 failed, 152 passed, 95145 warnings in 952.64s (0:15:52) ---
--- PYTEST VET TESTS FAILED  17 failed, 50 passed, 490 warnings in 4439.25s (1:13:59) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 31/259 TESTS  ---

	modified:   pypeit/spectrographs/apf_levy.py
	modified:   pypeit/spectrographs/keck_deimos.py
	modified:   pypeit/spectrographs/keck_lris.py
	modified:   pypeit/spectrographs/keck_mosfire.py
	modified:   pypeit/spectrographs/shane_kast.py

tbowers7 committed:
Merge branch 'develop' into issue_1958

tbowers7 committed:
More modifications to deal with test failures.

In concert with the updated ``.pypeit`` files in the DevSuite, this
*should* fix everything.  🤞

	modified:   pypeit/core/meta.py
	modified:   pypeit/spectrographs/gemini_gmos.py
	modified:   pypeit/spectrographs/gtc_osiris.py
	modified:   pypeit/spectrographs/keck_mosfire.py
	modified:   pypeit/spectrographs/ldt_deveny.py

tbowers7 committed:
Merge branch 'develop' into issue_1958

tbowers7 commented:
```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  261 passed, 441 warnings in 414.78s (0:06:54) ---
--- PYTEST UNIT TESTS FAILED  5 failed, 152 passed, 95145 warnings in 952.64s (0:15:52) ---
--- PYTEST VET TESTS FAILED  17 failed, 50 passed, 490 warnings in 4439.25s (1:13:59) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 31/259 TESTS  ---
Failed tests:
...
Skipped tests:
...
Testing Started at 2025-08-28T20:57:00.038527
Testing Completed at 2025-08-29T07:19:40.806523
Total Time: 10:22:40.767996
```

Only slightly more successful than the first test.  The biggest issue here was the fact that I needed to update the canned ``.pypeit`` files in the DevSuite (https://github.com/pypeit/PypeIt-development-suite/pull/369) to include the new columns written to satisfy ``config_specific_par()`` when passed a metadata table row.

Here's hoping! 🤞

tbowers7 committed:
Pull in recent LBT/MODS changes

	modified:   pypeit/spectrographs/lbt_mods.py

tbowers7 committed:
Merge branch 'develop' into issue_1958

tbowers7 committed:
Merge remote-tracking branch 'pypeit/develop' into issue_1958

tbowers7 committed:
Fixing issues related to latest test failures

Several spectrograph-specific issues, some related to reecnt PR's that
were just merged in.

For the InputFiles, this is the crux of the issue.  Only the PypeIt
file now sends a Table row to the spectrograph's ``config_specific_par()``
method, since this is where users will correct errant FITS header
information to create a Big Beautiful PypeIt File.

All other input files (Coadd2D, etc.) pull information from PypeIt-
produced files, which should already contain the proper metadata
information.  Hmmm, actually, maybe it doesn't.  I need to think about
this a bit more...

	modified:   pypeit/inputfiles.py
	modified:   pypeit/spectrographs/gemini_gmos.py
	modified:   pypeit/spectrographs/gemini_gnirs.py
	modified:   pypeit/spectrographs/keck_hires.py
	modified:   pypeit/spectrographs/lbt_mods.py

tbowers7 committed:
``get_meta_value()`` is now smarter

When parsing metadata for FITS headers of PypeIt-generated files (e.g.,
spec2d_*.fits), ``get_meta_value()`` no longer tries to reconstruct
metadata from the RAW HEADER CARDS included in the header, but instead
pulls the metadata values direct from the FITS keywords written.  For
instance, if a spectrograph constructs a metadata value from a compound
statement (i.e., ``compound_meta()``), the constructed value is already
written to the downstream file header.

This also means that I can probably remove the ``raw_header_cards()``
hack I added a couple years ago, since this is the exact problem I was
trying to solve then.  The idea was I needed to generate
``config_specific_par()`` from downstream files, and I thought at the
time that including the raw keys needed to reconstruct those metadata
was the way to go.  Unfortunately, I didn't consider the effects of bad
FITS keywords in the first place.  This should fix everything, I think.

	modified:   pypeit/spectrographs/spectrograph.py

tbowers7 commented:
```Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  261 passed, 441 warnings in 328.83s (0:05:28) ---
--- PYTEST UNIT TESTS PASSED  157 passed, 95113 warnings in 826.42s (0:13:46) ---
--- PYTEST VET TESTS FAILED  7 failed, 60 passed, 1188 warnings in 4066.32s (1:07:46) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 11/277 TESTS  ---
Failed tests:
    keck_mosfire/K_long pypeit
    keck_deimos/830G_M_9000_dither pypeit_coadd_2dspec
    gemini_gmos/GN_HAM_NS_B600_620 pypeit
    keck_mosfire/Y_long pypeit_ql
    keck_lris_blue/multi_600_4000_d560 pypeit_coadd_2dspec
    keck_mosfire/long2pos1_H pypeit_coadd_2dspec
    keck_lris_red_mark4/long_600_10000_d680 pypeit_ql
    bok_bc/300 pypeit
    gemini_gnirs_ifu/LR_IFU_32mm pypeit
    gemini_gmos/GS_HAM_B600_MOS pypeit
    gemini_gnirs_echelle/32_SB_SXD pypeit_coadd_2dspec
Skipped tests:
    gemini_gnirs_echelle/32_SB_SXD pypeit_tellfit
Testing Started at 2025-09-03T22:35:43.244475
Testing Completed at 2025-09-04T10:05:53.528349
Total Time: 11:30:10.283874
```

Getting closer.  The most recent commits above _should_ address the remainder of the failures.  There were a couple spectrograph-specific issues, but all the ``2dspec`` and ``ql`` failures were related to how I was pulling metadata for these PypeIt-produced files.  That should be fixed now.  🤞  Re-running tests (again).

tbowers7 commented:
```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  261 passed, 443 warnings in 507.78s (0:08:27) ---
--- PYTEST UNIT TESTS PASSED  157 passed, 93133 warnings in 1066.54s (0:17:46) ---
--- PYTEST VET TESTS FAILED  5 failed, 63 passed, 1609 warnings in 823.59s (0:13:43) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 7/280 TESTS  ---
Failed tests:
    gemini_gmos/GN_HAM_NS_B600_620 pypeit
    keck_mosfire/long2pos1_H pypeit_coadd_2dspec
    bok_bc/300 pypeit
    keck_mosfire/Y_long pypeit_ql
    keck_nires/ABBA_nostandard_faint pypeit_coadd_2dspec
    gemini_gnirs_ifu/LR_IFU_32mm pypeit
    gemini_gmos/GS_HAM_B600_MOS pypeit
Skipped tests:
Testing Started at 2025-09-05T17:29:46.681029
Testing Completed at 2025-09-06T11:58:31.586059
Total Time: 18:28:44.905030
```
Close, but still need to deal with some spectrograph-specific failures, including ones that shouldn't fail given the lack of substantive changes to the spectrograph class (like Bok/B&C).  Weird that Keck/NIRES now fails, since that spectrograph class was last touched over a year ago.

All failed VET tests can be traced back to one or another of the failed DevSuite tests (either same issue or missing output files).

tbowers7 committed:
Another round of whack-a-mole

Fixing remaining issues which *should* get us to passing all tests.

tbowers7 committed:
Merge remote-tracking branch 'pypeit/release' into issue_1958

tbowers7 commented:
Re-running tests (again again). 🤞

tbowers7 commented:
```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  261 passed, 443 warnings in 516.69s (0:08:36) ---
--- PYTEST UNIT TESTS FAILED  1 failed, 157 passed, 93130 warnings in 998.64s (0:16:38) ---
--- PYTEST VET TESTS FAILED  3 failed, 65 passed, 1605 warnings in 801.05s (0:13:21) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 2/280 TESTS  ---
Failed tests:
    bok_bc/300 pypeit
    gemini_gmos/GS_HAM_B600_MOS pypeit
Skipped tests:
Testing Started at 2025-09-10T21:58:54.769352
Testing Completed at 2025-09-11T16:40:41.772757
Total Time: 18:41:47.003405
```
Closer.  Feels like being stuck in Zeno's Paradox.  (Or is it Zeno's Paradise? 🏝️)  The failed DevSuite unit test is my new one.  🤦‍♂️  Paths issue.  2/3 VET test failures follow directly from the failed DevSuite tests.  The other VET test is ``test_coadd_datacube``, where a FITS keyword that should be present from PypeIt metadata is missing.

The sources of the ``bok_bc`` and ``gemini_gmos`` failures are each a bit mysterious at this point.  More investigation is required.

tbowers7 committed:
More test-failure "whack-a-mole"

	modified:   pypeit/spectrographs/gemini_gnirs.py
	modified:   pypeit/spectrographs/gtc_osiris.py
	modified:   pypeit/spectrographs/keck_kcwi.py
	modified:   pypeit/spectrographs/keck_nires.py

tbowers7 committed:
Python and its dynamic typing...

	modified:   pypeit/spectrographs/bok_bc.py

tbowers7 commented:
Starting the DevSuite tests once again.  🤞

The last two issues were stinkers.  The first was Python's dynamic typing and Bok/B&C not following the usual scheme of "300GPM" or "300L" for grating names... just "300" -- ``int(300)`` is not the same as ``str(300)``.  The second was older Gemini/GMOS ``.pypeit`` files using observing epoch in the MJD metadata column... which makes for Bad Things when trying to use that value as an actual MJD...

tbowers7 committed:
Add 'parangle' to outputfile for coadd3d's

Also, catch error earlier about ``None`` slits that made me scratch my
head for a while.  Now emits a sensible error message in ``edgetrace.py``
rather than crashing on some nonesense about ``None`` and ``int`` being
not compatable.

	modified:   pypeit/edgetrace.py
	modified:   pypeit/spectrographs/gemini_gnirs.py
	modified:   pypeit/spectrographs/gtc_osiris.py
	modified:   pypeit/spectrographs/keck_kcwi.py

tbowers7 commented:
🎉Zeno's Advancement Achieved. 🎉  Whack-a-Mole Mode Unlocked!
```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  261 passed, 443 warnings in 600.95s (0:10:00) ---
--- PYTEST UNIT TESTS PASSED  158 passed, 93126 warnings in 1096.45s (0:18:16) ---
--- PYTEST VET TESTS FAILED  3 failed, 65 passed, 1612 warnings in 966.32s (0:16:06) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 1/278 TESTS  ---
Failed tests:
    gemini_gmos/GS_HAM_R400_700 pypeit_flux
Skipped tests:
    gemini_gmos/GS_HAM_R400_700 pypeit_coadd_1dspec
    gemini_gmos/GS_HAM_R400_700 pypeit_tellfit
Testing Started at 2025-09-13T01:43:36.705572
Testing Completed at 2025-09-14T00:21:25.523093
Total Time: 22:37:48.817521
```
Remember the MJD issue?  Yeah, it caused these test failures because the filenames in the ``.flux`` and ``.coadd1d`` files are based on _when_ the object was observed... which is computed from the MJD.  🤣

tbowers7 committed:
SlicerIFU spectrograph downstream keywords

With the shift to relying on PypeIt metadata rather than raw-file
FITS headers, we need to include additional DAR-related metadata
keywords in downstream files for SlicerIFU spectrographs.  Placing this
specification in the ``subheader_for_spec()`` method means that all
future SlicerIFU spectrographs will automatically have these keywords
included in the downstream files.

	modified:   pypeit/spectrographs/gemini_gnirs.py
	modified:   pypeit/spectrographs/gtc_osiris.py
	modified:   pypeit/spectrographs/keck_kcwi.py
	modified:   pypeit/spectrographs/spectrograph.py

tbowers7 commented:
DevSuite tests pass (``tbowers-iss1958-15sep.report``).  Ready for review.
```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  261 passed, 441 warnings in 360.11s (0:06:00) ---
--- PYTEST UNIT TESTS PASSED  158 passed, 95104 warnings in 835.52s (0:13:55) ---
--- PYTEST VET TESTS PASSED  68 passed, 1613 warnings in 4316.17s (1:11:56) ---
--- PYPEIT DEVELOPMENT SUITE PASSED 280/280 TESTS  ---
Testing Started at 2025-09-15T23:32:00.336751
Testing Completed at 2025-09-16T12:19:35.188730
Total Time: 12:47:34.851979
```

tbowers7 committed:
Merge branch 'develop' into issue_1958

tbowers7 committed:
Merge branch 'develop' into issue_1958

tbowers7 commented:
Based on conversation during the PypeIt Developers Meeting on 22 September 2025, will work to make this PR backwards-compatible.  Marking as draft for now.  Review of changes to this point are encouraged.

tbowers7 committed:
Merge branch 'develop' into issue_1958

tbowers7 committed:
Merge branch 'develop' into issue_1958

tbowers7 committed:
Merge branch 'develop' into issue_1958

tbowers7 committed:
First pass at backwards compatability

	modified:   pypeit/spectrographs/ldt_deveny.py
	modified:   pypeit/spectrographs/spectrograph.py

tbowers7 committed:
Apply backwards-compatability change to all spectrographs

In the event that we are using an older ``.pypeit`` file for a
spectrograph that needs additional columns in the metadata, be able
to pull the relevant information from the FITS header by sending
to the ``get_meta_value()`` function the full path to the raw frame.

Propagate these changes to all ``config_specifi_par()`` functions
in all spectrographs.  This does have the effect of cleaning up
the individual spectrograph methods a little.

	modified:   pypeit/inputfiles.py
	modified:   pypeit/spectrographs/aat_uhrf.py
	modified:   pypeit/spectrographs/apf_levy.py
	modified:   pypeit/spectrographs/bok_bc.py
	modified:   pypeit/spectrographs/gemini_flamingos.py
	modified:   pypeit/spectrographs/gemini_gmos.py
	modified:   pypeit/spectrographs/gemini_gnirs.py
	modified:   pypeit/spectrographs/gtc_osiris.py
	modified:   pypeit/spectrographs/keck_deimos.py
	modified:   pypeit/spectrographs/keck_hires.py
	modified:   pypeit/spectrographs/keck_kcwi.py
	modified:   pypeit/spectrographs/keck_lris.py
	modified:   pypeit/spectrographs/keck_mosfire.py
	modified:   pypeit/spectrographs/keck_nirspec.py
	modified:   pypeit/spectrographs/lbt_luci.py
	modified:   pypeit/spectrographs/lbt_mods.py
	modified:   pypeit/spectrographs/ldt_deveny.py
	modified:   pypeit/spectrographs/mmt_binospec.py
	modified:   pypeit/spectrographs/mmt_bluechannel.py
	modified:   pypeit/spectrographs/mmt_mmirs.py
	modified:   pypeit/spectrographs/not_alfosc.py
	modified:   pypeit/spectrographs/ntt_efosc2.py
	modified:   pypeit/spectrographs/p200_dbsp.py
	modified:   pypeit/spectrographs/shane_kast.py
	modified:   pypeit/spectrographs/soar_goodman.py
	modified:   pypeit/spectrographs/spectrograph.py
	modified:   pypeit/spectrographs/subaru_focas.py
	modified:   pypeit/spectrographs/tng_dolores.py
	modified:   pypeit/spectrographs/vlt_fors.py
	modified:   pypeit/spectrographs/wht_isis.py

tbowers7 commented:
This PR is now backwards compatible.  Running DevSuite tests now -- will post results when complete.

tbowers7 committed:
One little line, so much sadness

Who woulda thunk that this one little line would cause so much
devastation in the DevSuite tests?

Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  276 passed, 471 warnings in 499.74s (0:08:19) ---
--- PYTEST UNIT TESTS PASSED  158 passed, 93527 warnings in 1079.05s (0:17:59) ---
--- PYTEST VET TESTS FAILED  15 failed, 54 passed, 1613 warnings in 6882.61s (1:54:42) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 34/276 TESTS  ---
Failed tests:
    keck_lris_blue/multi_300_5000_d680 pypeit
    keck_lris_red_mark4/long_400_8500_d560 pypeit
    keck_lris_blue_orig/long_600_4000_d500 pypeit
    keck_lris_red/long_600_7500_d560 pypeit
    keck_lris_red/multi_1200_9000_d680_1x2 pypeit
    keck_lris_red/multi_1200_9000_d680 pypeit
    keck_lris_red/multi_600_5000_d560 pypeit
    keck_lris_red/multi_400_8500_d560 pypeit
    keck_lris_red/long_600_10000_d680 pypeit
    keck_lris_red/long_400_8500_longread pypeit
    keck_lris_blue/multi_600_4000_slitmask pypeit
    keck_lris_blue/multi_600_4000_d560 pypeit
    keck_lris_red_orig/long_300_5000 pypeit
    keck_lris_red_mark4/long_600_10000_d680 pypeit
    keck_lris_blue/long_400_3400_d560 pypeit
    keck_lris_blue/long_600_4000_d560 pypeit
    keck_lris_red/mulit_831_8200_d560 pypeit
    keck_lris_red_orig/long_150_7500_d500 pypeit
    keck_lris_red_mark4/multi_600_10000_slitmask pypeit
    keck_lris_red_orig/long_400_8500_d560 pypeit
    keck_lris_red_orig/long_900_5500_d560 pypeit
    keck_lris_red_orig/long_831_8200_d460 pypeit
    keck_lris_red/long_1200_7500_d560 pypeit
    keck_lris_blue/long_600_4000_d560_slitless pypeit
    keck_lris_red/long_300_5000_d560 pypeit
    keck_lris_blue_orig/multi_1200_3400_d460 pypeit
    keck_lris_red/multi_900_5500_d560 pypeit
    keck_lris_red_orig/long_600_7500_d680 pypeit
    keck_lris_red_orig/long_1200_7500_d560 pypeit
    keck_lris_blue/multi_600_4000_d560_slitless pypeit
    keck_lris_red_orig/multi_600_5000_d500 pypeit
    keck_lris_red_mark4/multi_1200_9000_d680 pypeit
    keck_lris_red_orig/long_600_10000_d460 pypeit
    keck_lris_red/long_150_7500_d560 pypeit
Skipped tests:
    keck_lris_red/long_600_7500_d560 pypeit (ignore calibrations)
    keck_lris_red/long_600_7500_d560 pypeit_ql
    keck_lris_blue/multi_600_4000_d560 pypeit_coadd_2dspec
    keck_lris_red_mark4/long_600_10000_d680 pypeit_sensfunc
    keck_lris_red_mark4/long_600_10000_d680 pypeit_ql
Testing Started at 2025-10-10T20:56:13.573989
Testing Completed at 2025-10-11T15:04:03.555879
Total Time: 18:07:49.981890

	modified:   pypeit/spectrographs/keck_lris.py

tbowers7 commented:
DevSuite tests pass (`tbowers-iss1958-11oct.report`).  Now (finally), ready for backwards-compitible review!

The companion DevSuite PR (https://github.com/pypeit/PypeIt-development-suite/pull/369) no longer adds the new columns to the ``.pypeit`` files in that repo that would be written for new files under this PR, and thus serves as the test of backwards-compatibility for these changes.

```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  276 passed, 465 warnings in 329.96s (0:05:29) ---
--- PYTEST UNIT TESTS PASSED  158 passed, 93617 warnings in 811.74s (0:13:31) ---
--- PYTEST VET TESTS PASSED  69 passed, 153110 warnings in 4647.60s (1:17:27) ---
--- PYPEIT DEVELOPMENT SUITE PASSED 281/281 TESTS  ---
Testing Started at 2025-10-11T19:53:14.487458
Testing Completed at 2025-10-12T07:51:16.563706
Total Time: 11:58:02.076248
```

tbowers7 committed:
Release notes

	modified:   doc/releases/1.18.2dev.rst

kbwestfall commented on pypeit/scripts/chk_for_calibs.py:
```diff
@@ -44,20 +44,19 @@ def main(args):
 
-        from IPython import embed
-
+        import astropy.table
```
I'm throwing stones in glass houses on this one, but is there a specific reason you prefer this over `from astropy import table`?  Same for `pathlib` above.  If we keep the `import pathlib` line, we should remove the `from pathlib import Path` line above it.

tbowers7 replied:
When I'm reading through code, it helps me a lot to know exactly where a particular function comes from without having to scroll up to the import section.  I know it makes individual lines of code longer.  Now, for common things like `astropy.table.Table`, maybe I'm a bit overboard.  But I find the full import path for things in, _e.g._, scipy (`scipy.interpolate.curve_fit()`) to be more clear.

I realize this is a personal preference issue and I shouldn't be remaking a large project like PypeIt into my own image, but this is something I do in my own projects.  It also makes the import list _feel_ cleaner... I don't know what performance issues there are in importing one thing from a namespace rather than the whole namespace.

As I noted in a commit in #2013, I really need to stick to the point with these PRs and not try to make the code "look pretty" for my sole benefit.  I will revert these changes.

kbwestfall commented on pypeit/scripts/chk_for_calibs.py:
```diff
@@ -128,9 +127,12 @@ def main(args):
+            csf_idx = ps.fitstbl['filename'] == pathlib.Path(config_specific_file).name
+            spectrograph_cfg_lines = (
+                ps.spectrograph.config_specific_par(ps.fitstbl[csf_idx]).to_config()
+            )
```
Can we make this a method in `pypeit.metadata.PypeItMetaData`?

tbowers7 replied:
Yes, and so made.

kbwestfall commented on pypeit/spectrographs/aat_uhrf.py:
```diff
@@ -3,20 +3,21 @@
 
+import astropy.io.fits
+import astropy.table
+import astropy.time
```
Similar comment/question about the imports.

tbowers7 replied:
I'll revert.

tbowers7 replied:
Although, it would be helpful to have some PypeIt-standard way of doing these.  In some places it's`from astropy import table` and in others it's `from astropy.table import Table`.  I guess my attempt at `import astropy.table` was also an attempt to standardize.

kbwestfall replied:
FWIW, this doesn't bother me.  E.g., if I know I only need `Table` in a file, I'll `from astropy.table import Table`.  Other times I know I'll need both `Table` and `Column` and so I'll `from astropy import table` and use, e.g., `table.Table` and `table.Column`.  That doesn't feel inconsistent to me.  You do save memory by only importing what you need, but I understand the argument about consistency across the repository.  If we want to impose a standard, we should probably implement a pre-commit linter like astropy does, but I'm not sure it's worth the effort.  IMO, it's better to be flexible and limit barriers to new developers than it is to be too strict about standards (within some limits, of course).

tbowers7 replied:
That sounds reasonable.  I've reverted my superfluous edits, but stuck to `from astropy.time import Time` for all of the spectrograph classes.

tbowers7 committed:
Merge branch 'develop' into issue_1958

tbowers7 committed:
Merge branch 'develop' into issue_1958

tbowers7 committed:
Merge branch 'develop' into issue_1958

tbowers7 committed:
Merge branch 'develop' into issue_1958

kbwestfall commented on pypeit/spectrographs/spectrograph.py:
```diff
@@ -187,15 +189,38 @@ def default_pypeit_par(cls):
+    def config_specific_par(
+            self,
+            inp:str|list|pathlib.Path|astropy.io.fits.Header|astropy.table.Table,
+            inp_par:parset.ParSet=None
```
For typing, we should get into the habit of including `| None` for these kind of optional keywords.  I.e.
```
inp_par: parset.ParSet | None = None
```

tbowers7 replied:
Okay.  Didn't know that was a thing.

tbowers7 replied:
This particular parameter has been updated throughout the PR... still need to look for others.

kbwestfall commented on pypeit/spectrographs/spectrograph.py:
```diff
@@ -187,15 +189,38 @@ def default_pypeit_par(cls):
         specific instrument configurations.
 
+        This method may be called with either a row from the metadata table
+        (usually from the PypeIt Reduction File) or a filename/Header (used
```
Does this mean the type of `inp` can also be `astropy.table.row.Row`?

tbowers7 replied:
Oops, it's actually a single-row table.  I missed that when cleaning up the docstrings.

kbwestfall commented on pypeit/spectrographs/spectrograph.py:
```diff
@@ -187,15 +189,38 @@ def default_pypeit_par(cls):
+
+            # Adjust parameters based on settings used
+            grating = self.get_meta_value(scifile, 'dispname')
+            binning = self.get_meta_value(scifile, 'binning')
```
Do all spectrographs need both of these?  And should this be `scifile` or `inp`?

tbowers7 replied:
Good catch.  Not all spectrographs need these particular ones, this was an attempt at a template for new spectrographs that need `config_specific_pars()`.  And, yes, it should be `inp` instead of `scifile`.

tbowers7 replied:
I've add clarifying language to the docstring that `grating` and `binning` are shown as examples, but that the particular metadata needed for a specific instance of this method may be different.

kbwestfall commented on pypeit/spectrographs/spectrograph.py:
```diff
@@ -1411,14 +1452,27 @@ def get_meta_value(self, inp, meta_key, required=False,
+        # Otherwise, this set of statements pulls the Header array for the file in question
         else:
-            msgs.error(f'Unrecognized type for input: {type(inp)}')
+            if isinstance(inp, (str, pathlib.Path, astropy.io.fits.HDUList)):
```
Does the rest of this need to be nested in the `else` block?  Could you instead:
```
if instance(inp, astropy.table.Table):
    ...
elif isinstance(inp, (str, pathlib.Path, astropy.io.fits.HDUList):
    ...
...
```

tbowers7 replied:
You are correct.

kbwestfall commented on pypeit/spectrographs/spectrograph.py:
```diff
@@ -1438,6 +1492,12 @@ def get_meta_value(self, inp, meta_key, required=False,
+        #   then we KNOW that the desired metadata key exists as a FITS keyword already.
+        #   Return it now and skip all the gynmastics
+        if headarr[0].get('PIPELINE','').strip() == 'PYPEIT':
+            return headarr[0][meta_key.upper()]
```
Nice.  This feels like a useful shortcut.

kbwestfall commented on pypeit/spectrographs/gemini_flamingos.py:
```diff
@@ -164,15 +174,21 @@ def config_specific_par(self, scifile, inp_par=None):
+        par = super().config_specific_par(inp, inp_par=inp_par)
+
+        # Adjust parameters based on disperser used
+        grating = self.get_meta_value(inp, 'dispname')
```
Thanks for making it so that we're only doing this once!

kbwestfall commented on pypeit/spectrographs/gemini_gnirs.py:
```diff
@@ -829,13 +862,18 @@ def get_datacube_bins(self, slitlength, minmax, num_wave):
         return xbins, ybins, spec_bins
 
-    def pypeit_file_keys(self):
+    def configuration_keys(self):
```
Just checking that this is right.  I.e., that you removed the `pypeit_file_keys` function and added the `configuration_keys` function.

tbowers7 replied:
I did mean to do that... although after several iterations of how to make this work, I probably don't anymore.  Thanks.

tbowers7 replied:
This has been reverted to `pypeit_file_keys()`.

kbwestfall commented on pypeit/spectrographs/gtc_osiris.py:
```diff
@@ -328,78 +349,83 @@ def config_specific_par(self, scifile, inp_par=None):
+
+        # Adjust parameters based on grating and idname
+        grating = self.get_meta_value(inp, 'dispname')
+        idname = self.get_meta_value(inp, 'idname')
```
Not really related to this PR, but we're not using `idname` correctly here.  According to the definition in `meta.py`, this should be the observatory supplied frame type, not the observing mode.  I imaging there are other cases like this...

tbowers7 replied:
For better or worse, this FITS keyword (`OBSMODE`) is actually the closest thing the OSIRIS headers have for what we use as `idname`.  The other options are `OBSCLASS` which has two options: "Calib" and "Science", and `OBSTYPE` which is always "Object" for all frames.

So, while it also includes information about observing mode, it really is the frame type, also.  We may need to boot this one back to this instrument's cognizant individual.

kbwestfall commented on pypeit/spectrographs/keck_deimos.py:
```diff
@@ -9,35 +9,35 @@
 from pypeit.spectrographs import slitmask 
 from pypeit.spectrographs.opticalmodel import ReflectionGrating, OpticalModel, DetectorMap
 
+import linetools
```
Wow.  I guess this means that the `load_wmko_std_spectrum` is never used.  Can we remove it?

tbowers7 replied:
When I do a "find all references" for `load_wmko_std_spectrum`, the only hit is the function definition itself.  Yes, I will remove this function.

kbwestfall replied:
I just thought to check the dev-suite, and it is in `unit_tests/test_wmko_flux_std.py`.  I don't know how that has been passing if `linetools` was unavailable!  Anyway, we should probably leave it.  Sorry for the confusion!

tbowers7 replied:
`linetools` was imported... I had moved it around in my "messing".  It is back to its original position now.  If `load_wmko_std_spectrum` is not called anywhere in the main repo, and only appears in the one DevSuite `unit_test`, it begs the question of whether we need it or not at all (_i.e._, remove both the function here and its `unit_test`)?

kbwestfall replied:
Yeah, we should probably just get rid of both, but let's punt that until later.

kbwestfall commented on pypeit/spectrographs/keck_lris.py:
```diff
@@ -756,14 +768,14 @@ def get_maskdef_slitedges(self, ccdnum=None, filename=None, debug=None,
 
 
-        hdu = fits.open(filename)
+        hdu = astropy.io.fits.open(filename)
```
We probably should be using `pypeit.io.fits_open` instead...

tbowers7 replied:
Changed the 3 examples of `fits.open()` found in the spectrograph classes.

kbwestfall commented on pypeit/spectrographs/keck_nires.py:
```diff
@@ -431,9 +431,7 @@ def pypeit_file_keys(self):
-        pypeit_keys = super().pypeit_file_keys()
-        pypeit_keys += ['dithpat', 'dithpos', 'dithoff','frameno']
-        return pypeit_keys
+        return super().pypeit_file_keys() + ['dithpat', 'dithpos', 'dithoff', 'frameno']
```
Nice.

kbwestfall commented on pypeit/spectrographs/ldt_deveny.py:
```diff
@@ -425,9 +428,12 @@ def pypeit_file_keys(self):
         """
-        return super().pypeit_file_keys() + ['dispangle','slitwid','lampstat01']
+        pypeit_keys = super().pypeit_file_keys()
+        pypeit_keys.remove('decker')  # Until the decker is actually installed
```
To future proof this, you might want to put these statements in `if` or `try` blocks:
```
try:
    pypeit_keys.remove('decker')
except ValueError:
   pass
```
or
```
if 'decker' in pypeit_keys:
    pypeit_keys.remove('decker')
```

tbowers7 replied:
Good catch.

kbwestfall commented on pypeit/spectrographs/ldt_deveny.py:
```diff
@@ -469,93 +480,95 @@ def config_specific_par(self, scifile, inp_par=None):
+                par['sensfunc']['UVIS']['resolution'] = 3000
+
+            case 'DV10 (2160/5000)':
+                # Presently unsupported; no parameter changes
```
It it worth throwing a warning here?  (and below for the default case)?

tbowers7 replied:
Yes, thank you.

kbwestfall commented on pypeit/tests/test_spectrographs.py:
```diff
@@ -149,3 +149,42 @@ def test_atmext():
+    ft2 = fitstbl.table.copy()
+    ft2['dispname'][csf_idx] = '452/3306'
+    par = spectrograph.config_specific_par(ft2[csf_idx])
+    assert par['calibrations']['wavelengths']['reid_arxiv'] == 'shane_kast_blue_452.fits'
```
Nice!

kbwestfall commented on pypeit/edgetrace.py:
```diff
@@ -769,7 +769,7 @@ def auto_trace(self, bpm=None, debug=0):
             if debug > 0:
                 self.show(title='After matching to slit-mask design metadata.')
-            if np.all(self.bitmask.flagged(self.edge_msk, self.bitmask.bad_flags)):
+            if self.edge_msk is None or np.all(self.bitmask.flagged(self.edge_msk, self.bitmask.bad_flags)):
```
Good catch!

kbwestfall commented on pypeit/inputfiles.py:
```diff
@@ -698,24 +700,45 @@ def get_pypeitpar(self):
         # set of file names each time they are requested.  However, this should
         # only be done once in the code below because as soon as a relevant file
         # is found the loops are discontinued using `break`.
+        filenames = self.filenames.copy()
```
Good call to save this instead of using the `property` method every time.  But do we need the `copy`?

tbowers7 replied:
Good question.  The property does a disk-search to return the value... wasn't sure if just assigning the output to `filenames` would alleviate the disk-search.

tbowers7 replied:
I think we don't need the `copy()`.  Since the `@property` returns a list, we can do whatever we need to with it, but `filenames` is just a list.

kbwestfall reviewed (approved):
Looks good, and thanks for seeing this one though!  I think this is ready to go.  I'm interested in two thoughts that might spark additional efforts:

- Personally, I'm not keen on the change to imports that force importing the full namespace sequence.  E.g., you convert `from astropy.io import fits` to `import astropy.io.fits`.  As I said in one of the comments, I'm throwing stones in glass houses on this one, but I'm interested to know the motivation.
- Some of this made me think we need a "SpectrographConfiguration" class that becomes an attribute of `Spectrograph`.  We effectively keep these in `fitstbl` now, but passing this to `Spectroraph` would keep us from having to do things like keep `self.dispname` (and similar) for some spectrographs.  Instead, we'd have something like `self.cfg` and then use that to access the necessary setup/config information.

tbowers7 committed:
Time flies like an arrow...

... fruit flies like a banana.

Reverting superfluous edits, particularly in the ordering and "from-ing"
of imports.  This should greatly clean up the PR so that the actual
meaningful changes proposed are more easily visible.  Also, it should be
left to the package maintainers about how to standardize import
statements.

	modified:   pypeit/scripts/chk_for_calibs.py
	modified:   pypeit/spectrographs/aat_uhrf.py
	modified:   pypeit/spectrographs/apf_levy.py
	modified:   pypeit/spectrographs/bok_bc.py
	modified:   pypeit/spectrographs/gemini_flamingos.py
	modified:   pypeit/spectrographs/gemini_gmos.py
	modified:   pypeit/spectrographs/gemini_gnirs.py
	modified:   pypeit/spectrographs/gtc_osiris.py
	modified:   pypeit/spectrographs/keck_deimos.py
	modified:   pypeit/spectrographs/keck_hires.py
	modified:   pypeit/spectrographs/keck_kcwi.py
	modified:   pypeit/spectrographs/keck_lris.py
	modified:   pypeit/spectrographs/keck_mosfire.py
	modified:   pypeit/spectrographs/keck_nirspec.py
	modified:   pypeit/spectrographs/lbt_luci.py
	modified:   pypeit/spectrographs/lbt_mods.py
	modified:   pypeit/spectrographs/mmt_binospec.py
	modified:   pypeit/spectrographs/mmt_bluechannel.py
	modified:   pypeit/spectrographs/mmt_mmirs.py
	modified:   pypeit/spectrographs/not_alfosc.py
	modified:   pypeit/spectrographs/ntt_efosc2.py
	modified:   pypeit/spectrographs/p200_dbsp.py
	modified:   pypeit/spectrographs/shane_kast.py
	modified:   pypeit/spectrographs/soar_goodman.py
	modified:   pypeit/spectrographs/spectrograph.py
	modified:   pypeit/spectrographs/subaru_focas.py
	modified:   pypeit/spectrographs/tng_dolores.py
	modified:   pypeit/spectrographs/vlt_fors.py
	modified:   pypeit/spectrographs/wht_isis.py

tbowers7 committed:
Merge branch 'develop' into issue_1958

tbowers7 committed:
PR Comments, round 1

These are the low-hanging fruit; more PR comment corrections coming.

	modified:   pypeit/spectrographs/gemini_gmos.py
	modified:   pypeit/spectrographs/keck_deimos.py
	modified:   pypeit/spectrographs/keck_lris.py
	modified:   pypeit/spectrographs/ldt_deveny.py
	modified:   pypeit/spectrographs/mmt_bluechannel.py
	modified:   pypeit/spectrographs/spectrograph.py

tbowers7 committed:
Exercises in typing

	modified:   pypeit/spectrographs/p200_dbsp.py
	modified:   pypeit/spectrographs/p200_ngps.py

tbowers7 commented:
> * Some of this made me think we need a "SpectrographConfiguration" class that becomes an attribute of `Spectrograph`.  We effectively keep these in `fitstbl` now, but passing this to `Spectroraph` would keep us from having to do things like keep `self.dispname` (and similar) for some spectrographs.  Instead, we'd have something like `self.cfg` and then use that to access the necessary setup/config information.

This idea lines up with #2014 , so maybe we include this thought there and think about how to implement a "SpectrographConfiguration" class.

tbowers7 committed:
Merge branch 'develop' into issue_1958

tbowers7 committed:
Merge branch 'develop' into issue_1958

debora-pe reviewed (approved):
This looks good!
Thanks @tbowers7

tbowers7 committed:
Metadata method for getting table row

And corresponding changes elsewhere.

	modified:   pypeit/metadata.py
	modified:   pypeit/scripts/chk_for_calibs.py
	modified:   pypeit/tests/test_metadata.py
	modified:   pypeit/tests/test_spectrographs.py

tbowers7 committed:
Clean keys in spectrographs

Going through the spectrographs after all of the ups and downs of this
PR and making sure all of the ``key`` edits conform to the final version
of how I did things.

This should also finish out all PR comments.

	modified:   pypeit/inputfiles.py
	modified:   pypeit/metadata.py
	modified:   pypeit/spectrographs/aat_uhrf.py
	modified:   pypeit/spectrographs/apf_levy.py
	modified:   pypeit/spectrographs/bok_bc.py
	modified:   pypeit/spectrographs/gemini_flamingos.py
	modified:   pypeit/spectrographs/gemini_gmos.py
	modified:   pypeit/spectrographs/gemini_gnirs.py
	modified:   pypeit/spectrographs/gtc_osiris.py
	modified:   pypeit/spectrographs/keck_deimos.py
	modified:   pypeit/spectrographs/keck_hires.py
	modified:   pypeit/spectrographs/keck_kcwi.py
	modified:   pypeit/spectrographs/keck_lris.py
	modified:   pypeit/spectrographs/keck_mosfire.py
	modified:   pypeit/spectrographs/keck_nirspec.py
	modified:   pypeit/spectrographs/lbt_luci.py
	modified:   pypeit/spectrographs/lbt_mods.py
	modified:   pypeit/spectrographs/ldt_deveny.py
	modified:   pypeit/spectrographs/mmt_binospec.py
	modified:   pypeit/spectrographs/mmt_bluechannel.py
	modified:   pypeit/spectrographs/mmt_mmirs.py
	modified:   pypeit/spectrographs/not_alfosc.py
	modified:   pypeit/spectrographs/ntt_efosc2.py
	modified:   pypeit/spectrographs/p200_dbsp.py
	modified:   pypeit/spectrographs/shane_kast.py
	modified:   pypeit/spectrographs/soar_goodman.py
	modified:   pypeit/spectrographs/spectrograph.py
	modified:   pypeit/spectrographs/subaru_focas.py
	modified:   pypeit/spectrographs/tng_dolores.py
	modified:   pypeit/spectrographs/vlt_fors.py
	modified:   pypeit/spectrographs/wht_isis.py

tbowers7 committed:
Merge branch 'develop' into issue_1958

tbowers7 commented:
Since I made some changes to configuration and pypeit_file keys, I am running the DevSuite ~~one more time~~ two more times to be sure everything is okay.  @kbwestfall, please have a look that your comments are adequately addressed.  I think we should add the idea of "SpectrographConfiguration" class to #2014.

tbowers7 committed:
Finesse the metadata method test

	modified:   pypeit/tests/test_metadata.py

tbowers7 committed:
Fix test failures

* Duplicate column for Keck/LRIS ('decker' is in core)
* Missed a ``.is_file()`` when converting ``os`` --> ``pathlib`` for Gemini/GMOS

	modified:   pypeit/spectrographs/gemini_gmos.py
	modified:   pypeit/spectrographs/keck_lris.py

tbowers7 commented:
DevSuite tests pass -- although it took a second pass over VLT/X-Shooter to be sure.  There was a weird error (enumerated in #2033) in the fitting of the NIR arm of X-Shooter in the first pass that didn't show up the second time.  (Please note that this PR does not touch X-Shooter's spectrograph class.)

First pass (`tbowers-iss1958-01nov.report`):
```
Test Summary
--------------------------------------------------------
--- PYTEST PYPEIT UNIT TESTS PASSED  278 passed, 471 warnings in 395.04s (0:06:35) ---
--- PYTEST UNIT TESTS PASSED  158 passed, 93527 warnings in 935.09s (0:15:35) ---
--- PYTEST VET TESTS FAILED  2 failed, 67 passed, 133816 warnings in 4996.56s (1:23:16) ---
--- PYPEIT DEVELOPMENT SUITE FAILED 1/281 TESTS  ---
Failed tests:
    vlt_xshooter/NIR pypeit
Skipped tests:
Testing Started at 2025-11-01T16:10:17.707513
Testing Completed at 2025-11-02T06:31:58.725825
Total Time: 14:21:41.018312
```
The two VET test failures were related to missing X-Shooter NIR output files.

When just VLT/X-Shooter was run, everything passed (`tbowers-iss1958-02nov.report`)
```
Test Summary
--------------------------------------------------------
--- PYPEIT DEVELOPMENT SUITE PASSED 17/17 TESTS  ---
Testing Started at 2025-11-02T17:02:55.115463
Testing Completed at 2025-11-02T17:53:17.452001
Total Time: 0:50:22.336538
```

So, I think this PR is ready for merge once @kbwestfall takes a second look.
